File: rust/src/bin/p0012.rs
// Project Euler: Problem 12

/*
- Assume that f(n) returns number of divisors of 'n'.
    f(a*b) = f(a) * f(b) when 'a' and 'b' are relatively prime.
- Triangle number's formula is n(n + 1)/2. Besides 'n' and 'n + 1' are relatively prime.

Therefore, ...
  - 'n/2' and 'n+1' are relatively prime (when 'n' is even)
  - 'n' and '(n+1)/2' are relatively prime (when 'n' is odd)
*/

euler::run_solver!(12);

fn solve() -> String {
    compute(500).to_string()
}

fn compute(thr: i64) -> i64 {
    let mut n: i64 = 1;
    loop {
        if num_of_divs(n) * num_of_divs((n + 1) / 2) > thr {
            break;
        };
        n += 1;
        if num_of_divs(n / 2) * num_of_divs(n + 1) > thr {
            break;
        }
        n += 1;
    }
    (n * (n + 1)) / 2
}

fn num_of_divs(n: i64) -> i64 {
    use euler::math;

    if n == 1 {
        1
    } else {
        math::factorize(n)
            .into_iter()
            .map(|(_, x)| x + 1)
            .product::<i64>()
    }
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0012_threshold_5() {
        assert_eq!(compute(5), 28);
    }

    #[test]
    fn p0012_threshold_500() {
        assert_eq!(compute(500), 76576500);
    }
}

File: ocaml/src/p0012.ml
(* Project Euler: Problem 12 *)

(*
   triangle number's formula is n(n + 1)/2 and 'n' and 'n + 1' are coprime.
   Therefore, ...
   - 'n/2' and 'n+1' are coprime (when 'n' is even)
   - 'n' and '(n+1)/2' are coprime (when 'n' is odd)

   assume that f(n) returns number of divisors of 'n'.
   f(a*b) = f(a) * f(b) when 'a' and 'b' are coprime.
*)
open Core

let num_of_divs n = Euler.Math.divisors n |> List.length

let find_num thr =
  let rec aux n =
    match n mod 2 with
    | 1 ->
      if num_of_divs n * num_of_divs ((n + 1) / 2) > thr
      then n * (n + 1) / 2
      else aux (succ n)
    | _ ->
      if num_of_divs (n / 2) * num_of_divs (n + 1) > thr
      then n * (n + 1) / 2
      else aux (succ n)
  in
  aux 1
;;

let compute thr = find_num thr
let solve () = compute 500 |> Int.to_string

(* Test *)

let%test_unit "5" = [%test_eq: int] (compute 5) 28
let%test_unit "500" = [%test_eq: int] (compute 500) 76576500
